src/Extensions/Compatibility/Validation/Code.cs

```diff
@@ -0,0 +1,515 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Omex.Extensions.Compatibility.Validation
+{
+	/// <summary>
+	/// Code validation
+	/// </summary>
+	public static class Code
+	{
+		/// <summary>
+		/// Check object's state, throws specified Exception type if state evaluates to false.
+		/// </summary>
+		/// <typeparam name="TException">Exception to throw</typeparam>
+		/// <param name="state">Object's state</param>
+		/// <param name="errorMessage">Error message</param>
+		/// <param name="tagId">Tag Id to log, leave null if no logging is needed</param>
+		[Obsolete(ObsoleteMessage, IsObsoleteError)]
+		public static void Expects<TException>(bool state, string errorMessage, EventId? tagId)
+			where TException : Exception
+		{
+			if (!Validate(state, errorMessage, tagId))
+			{
+				ReportError<TException>(errorMessage);
+			}
+		}
+
+
+		/// <summary>
+		/// Checks that the enumerable argument is not null and doesn't contain any nulls
+		/// </summary>
+		/// <remarks>Be careful to not pass enumerables that can be enumerated only once</remarks>
+		/// <typeparam name="TEnumerable">Type of the enumerable</typeparam>
```

Comment:
`TElement` might be more appropriate as it is the type of the element, rather than the type of the enumerable

Reply:
I don't want to change this class logic or spend time improving its documentation since it's marked as obsolete